### PR TITLE
Python: Fixes Issue 3206 where user input is echoed back as in AgentResponse when an agent emits user messages.

### PR DESCRIPTION
**Pull request by alliscode**
This pull request updates the agent workflow system to ensure that user-role messages are excluded from agent responses and streaming updates. The changes improve the correctness of agent outputs by filtering out user messages, so only assistant/tool-generated content is returned. Related tests have been updated and expanded to verify this behavior.

**Agent Output Filtering**

* Updated `_convert_workflow_events_to_agent_response` and related functions in `python/packages/core/agent_framework/_workflows/_agent.py` to filter out messages with the `user` role, ensuring only assistant/tool messages are included in agent responses and updates. [[1]](diffhunk://#diff-68f70a628740289bc19f66db9941df961c34b67715f7f7e0afa338d622472c54L478-R478) [[2]](diffhunk://#diff-68f70a628740289bc19f66db9941df961c34b67715f7f7e0afa338d622472c54R489-R494) [[3]](diffhunk://#diff-68f70a628740289bc19f66db9941df961c34b67715f7f7e0afa338d622472c54R583-R584) [[4]](diffhunk://#diff-68f70a628740289bc19f66db9941df961c34b67715f7f7e0afa338d622472c54R599-R600) [[5]](diffhunk://#diff-68f70a628740289bc19f66db9941df961c34b67715f7f7e0afa338d622472c54R617-R618)

**Test Updates and Additions**

* Modified tests in `python/packages/core/tests/workflow/test_workflow_agent.py` to assert that only non-user messages are present in agent responses and streaming updates, reflecting the new filtering logic. [[1]](diffhunk://#diff-c382584b0659384ba40a23b61dffbbc7c74c4d9f4f160198e2b25ef5b8479b0bL472-R476) [[2]](diffhunk://#diff-c382584b0659384ba40a23b61dffbbc7c74c4d9f4f160198e2b25ef5b8479b0bL494-R513)
* Added a new test in `python/packages/orchestrations/tests/test_handoff.py` to verify that user input is not echoed in streamed agent updates, ensuring compliance with the expected agent behavior. [[1]](diffhunk://#diff-8a2227bd8de9126d690d75cfba0dac19ef575e4b1cf22e28c34a35ec0074d547R11) [[2]](diffhunk://#diff-8a2227bd8de9126d690d75cfba0dac19ef575e4b1cf22e28c34a35ec0074d547R1121-R1156)

### Contribution Checklist

- [ ] The code builds clean without any errors or warnings
- [ ] The PR follows the [Contribution Guidelines](https://github.com/microsoft/agent-framework/blob/main/CONTRIBUTING.md)
- [ ] All unit tests pass, and I have added new tests where possible
- [ ] **Is this a breaking change?** If yes, add "[BREAKING]" prefix to the title of the PR.

### CODE REVIEW

**Comment by markwallace-microsoft** [comment]
<a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/README.md">Python Test Coverage</a><details><summary>Python Test Coverage Report • </summary><table><tr><th>File</th><th>Stmts</th><th>Miss</th><th>Cover</th><th>Missing</th></tr><tbody><tr><td colspan="5"><b>packages/core/agent_framework/_workflows</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py">_agent.py</a></td><td>386</td><td>75</td><td>80%</td><td><a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L66">66</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L74-L80">74&ndash;80</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L116-L117">116&ndash;117</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L207">207</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L253">253</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L255">255</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L318">318</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L320">320</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L382-L383">382&ndash;383</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L389-L390">389&ndash;390</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L396">396</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L398">398</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L403">403</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L464-L465">464&ndash;465</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L474">474</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L481">481</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L509">509</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L521">521</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L528">528</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L561-L563">561&ndash;563</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L565">565</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L567">567</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L569">569</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L574">574</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L579">579</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L617">617</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L632-L634">632&ndash;634</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L666">666</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L683">683</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L722-L725">722&ndash;725</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L731">731</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L737">737</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L741-L742">741&ndash;742</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L745-L751">745&ndash;751</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L755-L756">755&ndash;756</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L764">764</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L825">825</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L832">832</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L838-L839">838&ndash;839</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L850">850</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L882">882</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L889">889</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L910">910</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L919">919</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L923">923</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L925-L927">925&ndash;927</a>, <a href="https://github.com/microsoft/agent-framework/blob/642dc7ee41e454749c51cd91e76b968a62d51f99/packages/core/agent_framework/_workflows/_agent.py#L934">934</a></td></tr><tr><td><b>TOTAL</b></td><td><b>22634</b></td><td><b>2800</b></td><td><b>87%</b></td><td>&nbsp;</td></tr></tbody></table></details>

## Python Unit Test Overview
| Tests | Skipped | Failures | Errors | Time |
| ----- | ------- | -------- | -------- | ------------------ |
| 4704 | 25 :zzz: | 0 :x: | 0 :fire: | 1m 21s :stopwatch: |

File: python/packages/orchestrations/tests/test_handoff.py [hunk 1]
```diff
@@ -1117,3 +1118,39 @@ def get_session(self, *, service_session_id, **kwargs):
...
+        updates.append(update)
+
+    assert updates, "Expected at least one streaming update"
+
+    # The core assertion: no update should carry the user role.
+    user_role_updates = [u for u in updates if u.role == "user"]
+    assert not user_role_updates, (
+        f"User input was echoed back in the stream as {len(user_role_updates)} update(s). "
+        "Expected only assistant-role updates."
+    )
```
Comment (moonbox3) [inline python/packages/orchestrations/tests/test_handoff.py:1157]: This test only verifies the streaming path. The non-streaming path through `_convert_workflow_events_to_agent_response` is a separate code path that also received user-filtering changes. Add a non-streaming assertion to cover it.

```suggestion
    assert updates, "Expected at least one streaming update"

    # The core assertion: no update should carry the user role.
    user_role_updates = [u for u in updates if u.role == "user"]
    assert not user_role_updates, (
        f"User input was echoed back in the stream as {len(user_role_updates)} update(s). "
        "Expected only assistant-role updates."
    )

    # Also verify non-streaming path filters user messages
    result = await workflow_agent.run(user_input)
    user_role_messages = [m for m in result.messages if m.role == "user"]
    assert not user_role_messages, (
        f"User input was echoed back in non-streaming result as {len(user_role_messages)} message(s). "
        "Expected only assistant-role messages."
    )
```

**Review by moonbox3 — COMMENTED** [review]
## Automated Code Review

**Reviewers:** 3 | **Confidence:** 87%

### ✓ Correctness
> The diff consistently filters out user-role messages across all code paths (streaming and non-streaming) for every data type (AgentResponse, AgentResponseUpdate, Message, list[Message]). The logic is correct and the tests are properly updated to reflect the new filtering behavior. No bugs, race conditions, or missing error handling detected.

### ✗ Test Coverage
> The diff consistently filters user-role messages across four branches in both `_convert_workflow_events_to_agent_response` and `_convert_workflow_event_to_agent_response_updates`. The existing `test_workflow_as_agent_yield_output_with_list_of_chat_messages` test was updated to verify the `list[Message]` branch for both streaming and non-streaming paths, and a new integration test (`test_handoff_as_agent_run_stream_does_not_echo_user_input`) covers the handoff streaming scenario. However, several branches of the new filtering logic lack dedicated test coverage: the `AgentResponse` branch with mixed user/assistant messages, the single `Message` branch with role='user', the `AgentResponseUpdate` branch with role='user', and the edge case where all messages are user-role (producing an empty result). The new handoff test also only exercises the streaming path and omits the non-streaming `await agent.run()` verification.

### ✗ Design Approach
> The diff fixes user-message echoing in agent responses by sprinkling `role != "user"` guards across six sites in two conversion methods. While the symptom (handoff workflows leaking user input into streamed responses) is real, the fix is applied at the generic workflow-to-agent-response conversion layer rather than at the source (the handoff executor that emits the full conversation). This scatters a single filtering concern across many code paths, making it easy to miss a site if a new branch is added later, and silently drops user-role messages for all workflows—not just the handoff case that motivated the change. A centralised filter (or fixing the emitter) would be more robust. There is also an inconsistency: when an `AgentResponse` or `list[Message]` contains user messages that get filtered, the corresponding `raw_representations` entry is still appended unfiltered, whereas the single-`Message` branch skips the raw representation entirely for user messages.

### Flagged Issues
- [ ] No test covers filtering user messages from an `AgentResponse` containing a mix of user and assistant messages (the `isinstance(data, AgentResponse)` branch at line 478). This is a distinct code path from the `list[Message]` branch that is tested.
- [ ] No test covers the single `Message` with role='user' being dropped (the `isinstance(data, Message)` branch at lines 489-491 and 600-601). This path also skips appending to `raw_representations`, which differs from the other branches and should be verified.
- [ ] The user-message filter is scattered across six separate locations in two methods. If a new data-type branch is added to either method, forgetting the guard silently leaks user messages. Centralise the filter in one place—either a helper predicate applied once after collection, or better, prevent the upstream executor (HandoffAgentExecutor) from yielding user-role messages in the first place, keeping the generic conversion layer unaware of role semantics.
- [ ] Inconsistent handling of `raw_representations`: for the `AgentResponse` and `list[Message]` branches, raw representations are appended even when all messages were user-role (and thus filtered). For the single-`Message` branch, the raw representation is skipped. This means `raw_representations` can contain entries with no corresponding message, or vice-versa, depending on the code path.

### Suggestions
- In the `list[Message]` branch of `_convert_workflow_events_to_agent_response` (around line 496), `raw_representations.append(data)` still appends the full unfiltered list while `messages` only gets the filtered subset. This is arguably intentional (preserving raw data), but worth a brief comment to make the asymmetry explicit for future readers.
- Add a test where an executor yields a single `AgentResponseUpdate` with role='user' to verify the early return at line 575-576 produces an empty stream.
- Add an edge-case test where all yielded messages have role='user' to verify the agent returns an empty `AgentResponse.messages` list (or handles it gracefully).
- The new `test_handoff_as_agent_run_stream_does_not_echo_user_input` only tests `stream=True`. Add a non-streaming assertion (`result = await workflow_agent.run(user_input)`) to verify the non-streaming path also filters user messages, similar to how the updated `list[Message]` test covers both paths.
- Consider adding a single post-collection filter (e.g., `messages = [m for m in messages if m.role != "user"]`) at the end of `_convert_workflow_events_to_agent_response` and a similar wrapper in the update method, rather than guarding every branch independently.
- If user-message filtering is truly a universal invariant for agent responses, document it as such on `AgentResponse` (e.g., a class-level note or a validation in `__init__`), so future contributors don't re-introduce the issue.

---
*Automated review by moonbox3's agents*

File: python/packages/core/agent_framework/_workflows/_agent.py [hunk 2]
```diff
@@ -475,7 +475,7 @@ def _convert_workflow_events_to_agent_response(
                     )
 
                 if isinstance(data, AgentResponse):
-                    messages.extend(data.messages)
+                    messages.extend(msg for msg in data.messages if msg.role != "user")
```
Comment (moonbox3) [inline python/packages/core/agent_framework/_workflows/_agent.py]: This `AgentResponse` branch filters user messages from `data.messages` but still appends `data.raw_representation` unconditionally. This is a deliberate design choice that differs from the single `Message` branch (which skips `raw_representations` entirely for user messages). A unit test should verify this behavior — that an `AgentResponse` with mixed user/assistant messages produces filtered `messages` but retains the full `raw_representation`.

File: python/packages/core/agent_framework/_workflows/_agent.py [hunk 3]
```diff
@@ -572,13 +573,17 @@ def _convert_workflow_event_to_agent_response_updates(
 
             if isinstance(data, AgentResponseUpdate):
                 # Pass through AgentResponseUpdate directly (streaming from AgentExecutor)
+                if data.role == "user":
```
Comment (moonbox3) [inline python/packages/core/agent_framework/_workflows/_agent.py:604]: New early-return for `AgentResponseUpdate` with role='user' has no test coverage. An executor could emit an `AgentResponseUpdate` directly — add a test that yields one with role='user' and verifies it is dropped from the stream.

File: python/packages/core/agent_framework/_workflows/_agent.py (hunk 2 above)
Comment (moonbox3) [inline python/packages/core/agent_framework/_workflows/_agent.py]: This filters messages but the next line still appends `data.raw_representation` unconditionally. If all messages in `data` were user-role, you get a raw representation with no corresponding output message. The single-`Message` branch (line 489-491) skips `raw_representations` for user messages, creating an inconsistency.

File: python/packages/core/agent_framework/_workflows/_agent.py [hunk 4]
```diff
@@ -486,11 +486,12 @@ def _convert_workflow_events_to_agent_response(
...
+                    if data.role != "user":
+                        messages.append(data)
+                        raw_representations.append(data.raw_representation)
                 elif is_instance_of(data, list[Message]):
                     chat_messages = cast(list[Message], data)
-                    messages.extend(chat_messages)
+                    messages.extend(msg for msg in chat_messages if msg.role != "user")
                     raw_representations.append(data)
                 else:
                     contents = self._extract_contents(data)
```
Comment (moonbox3) [inline python/packages/core/agent_framework/_workflows/_agent.py:509]: Same issue: user messages are filtered from `messages` but `raw_representations.append(data)` on the next line still runs, even if every message in the list was user-role. Either filter `data` before appending to raw_representations, or centralise the filtering after collection so both lists stay consistent.

File: python/packages/core/agent_framework/_workflows/_agent.py (hunk 3 above)
Comment (moonbox3) [inline python/packages/core/agent_framework/_workflows/_agent.py:604]: Returning early for user-role `AgentResponseUpdate` silently swallows the event. If a future workflow intentionally emits a user-role update (e.g., for a 'thinking' UI that replays the user prompt), this will be very hard to debug. Consider filtering at the handoff layer instead of the generic conversion layer, or at minimum add a comment explaining why user-role updates are universally suppressed.

File: python/packages/core/agent_framework/_workflows/_agent.py (hunk 2 above)
Comment (TaoChenOSU) [inline python/packages/core/agent_framework/_workflows/_agent.py]: Why would the agent response contain user messages?

**Comment by TaoChenOSU** [comment]
I tried using this script to reproduce:
```python
import asyncio
import os
from typing import Annotated

from agent_framework import (
    Agent,
    AgentResponse,
    AgentResponseUpdate,
    Content,
    Message,
    WorkflowAgent,
    tool,
)
from agent_framework.azure import AzureOpenAIResponsesClient
from agent_framework.orchestrations import HandoffAgentUserRequest, HandoffBuilder
from azure.identity import AzureCliCredential
from dotenv import load_dotenv

load_dotenv()


@tool(approval_mode="never_require")
def process_refund(order_number: Annotated[str, "Order number to process refund for"]) -> str:
    return f"Refund processed successfully for order {order_number}."


@tool(approval_mode="never_require")
def check_order_status(order_number: Annotated[str, "Order number to check status for"]) -> str:
    return f"Order {order_number} is currently being processed and will ship in 2 business days."


@tool(approval_mode="never_require")
def process_return(order_number: Annotated[str, "Order number to process return for"]) -> str:
    return f"Return initiated successfully for order {order_number}. You will receive return instructions via email."


def create_agents(client: AzureOpenAIResponsesClient) -> tuple[Agent, Agent, Agent, Agent]:
    triage_agent = client.as_agent(
        instructions=(
            "You are frontline support triage. Route customer issues to the appropriate specialist agents "
            "based on the problem described."
        ),
        name="triage_agent",
    )

    refund_agent = client.as_agent(
        instructions="You process refund requests.",
        name="refund_agent",
        tools=[process_refund],
    )

    order_agent = client.as_agent(
        instructions="You handle order and shipping inquiries.",
        name="order_agent",
        tools=[check_order_status],
    )

    return_agent = client.as_agent(
        instructions="You manage product return requests.",
        name="return_agent",
        tools=[process_return],
    )

    return triage_agent, refund_agent, order_agent, return_agent


def handle_response_and_requests(response: AgentResponse) -> dict[str, HandoffAgentUserRequest]:
    pending_requests: dict[str, HandoffAgentUserRequest] = {}
    for message in response.messages:
        for content in message.contents:
            if content.type == "function_call":
                if isinstance(content.arguments, dict):
                    request = WorkflowAgent.RequestInfoFunctionArgs.from_dict(content.arguments)
                elif isinstance(content.arguments, str):
                    request = WorkflowAgent.RequestInfoFunctionArgs.from_json(content.arguments)
                else:
                    raise ValueError("Invalid arguments type. Expecting a request info structure for this sample.")
                if isinstance(request.data, HandoffAgentUserRequest):
                    pending_requests[request.request_id] = request.data

    return pending_requests


async def main() -> None:
    client = AzureOpenAIResponsesClient(
        project_endpoint=os.environ["AZURE_AI_PROJECT_ENDPOINT"],
        deployment_name=os.environ["AZURE_AI_MODEL_DEPLOYMENT_NAME"],
        credential=AzureCliCredential(),
    )

    triage, refund, order, support = create_agents(client)

    agent = (
        HandoffBuilder(
            name="customer_support_handoff",
            participants=[triage, refund, order, support],
            termination_condition=lambda conversation: (
                len(conversation) > 0 and "welcome" in conversation[-1].text.lower()
            ),
        )
        .with_start_agent(triage)
        .build()
        .as_agent()  # Convert workflow to agent interface
    )

    scripted_responses = [
        "My order 1234 arrived damaged and the packaging was destroyed. I'd like to return it.",
        "Please also process a refund for order 1234.",
        "Thanks for resolving this.",
    ]

    print("[Starting workflow with initial user message...]\n")
    initial_message = "Hello, I need assistance with my recent purchase."
    print(f"- User: {initial_message}")

    last_author: str | None = None
    chunks: list[AgentResponseUpdate] = []
    async for chunk in agent.run(initial_message, stream=True):
        chunks.append(chunk)
        if chunk.text:
            if chunk.author_name != last_author:
                print()  # New message, add a newline for readability
                last_author = chunk.author_name
                print(f"- {chunk.author_name}: ", end="", flush=True)
            print(chunk.text, end="", flush=True)

    pending_requests = handle_response_and_requests(AgentResponse.from_updates(chunks))

    while pending_requests:
        if not scripted_responses:
            responses = {req_id: HandoffAgentUserRequest.terminate() for req_id in pending_requests}
        else:
            user_response = scripted_responses.pop(0)
            print(f"\n- User: {user_response}")
            responses = {req_id: HandoffAgentUserRequest.create_response(user_response) for req_id in pending_requests}

        function_results = [
            Content.from_function_result(call_id=req_id, result=response) for req_id, response in responses.items()
        ]
        last_author: str | None = None
        chunks: list[AgentResponseUpdate] = []
        async for chunk in agent.run(Message("tool", function_results), stream=True):
            chunks.append(chunk)
            if chunk.text:
                if chunk.author_name != last_author:
                    print()  # New message, add a newline for readability
                    last_author = chunk.author_name
                    print(f"- {chunk.author_name}: ", end="", flush=True)
                print(chunk.text, end="", flush=True)

        pending_requests = handle_response_and_requests(AgentResponse.from_updates(chunks))


if __name__ == "__main__":
    asyncio.run(main())
```
(This is taken from https://github.com/microsoft/agent-framework/blob/main/python/samples/03-workflows/agents/handoff_workflow_as_agent.py and modified to run in streaming mode)

And here is the output I got:
```
[Starting workflow with initial user message...]

- User: Hello, I need assistance with my recent purchase.

- triage_agent: Could you elaborate on the issue with your recent purchase? Are you looking for help with a refund, return, or something else?
- User: My order 1234 arrived damaged and the packaging was destroyed. I'd like to return it.

- return_agent: The return for order 1234 has been initiated successfully. You'll receive further instructions via email on how to proceed with the return. If you have any more questions or need additional assistance, feel free to ask!
- User: Please also process a refund for order 1234.

- refund_agent: The refund for order 1234 has been processed successfully. If you need further assistance, feel free to let me know!
- User: Thanks for resolving this.

- refund_agent: You're welcome! If you have any more questions or need further assistance, feel free to reach out. Have a great day!Hello, I need assistance with my recent purchase.

- refund_agent: You're welcome! If you have any more questions or need further assistance, feel free to reach out. Have a great day!Hello, I need assistance with my recent purchase.
- triage_agent: Could you elaborate on the issue with your recent purchase? Are you looking for help with a refund, return, or something else?
- refund_agent: My order 1234 arrived damaged and the packaging was destroyed. I'd like to return it.
- return_agent: The return for order 1234 has been initiated successfully. You'll receive further instructions via email on how to proceed with the return. If you have any more questions or need additional assistance, feel free to ask!
- refund_agent: Please also process a refund for order 1234.The refund for order 1234 has been processed successfully. If you need further assistance, feel free to let me know!Thanks for resolving this.You're welcome! If you have any more questions or need further assistance, feel free to reach out. Have a great day!
```

I don't see the repeated user messages, but when the orchestration finishes it yields the full conversation (a list of messages) as the output, which gets converted to chunks each message. That's why we see the full conversation gets returned again. We need to fix that instead.

File: python/packages/core/agent_framework/_workflows/_agent.py (hunk 2 above)
Comment (alliscode) [reply python/packages/core/agent_framework/_workflows/_agent.py]: Because of things like our orchestrations that emit the full conversation.

**Comment by moonbox3** [comment]
Let's please revive this PR as I think it contains the proper direction towards the comprehensive fix (closing #4275 in favor of this). A few questions/concerns before we move forward:

1. `OrchestrationComplete` fallback doesn't filter. In the non-streaming path, when no other output events are collected, the fallback does `messages.extend(orchestration_complete.messages)` — the full conversation unfiltered. User messages would leak through in that case, which defeats the purpose.

2. Double-dipping on the fix. The PR introduces `OrchestrationComplete` to separate internal state from output, but also adds `role != "user"` filtering across every branch (`AgentResponse`, `Message`, `list[Message]`, `AgentResponseUpdate`). If `OrchestrationComplete` is doing its job, the role-based filtering shouldn't be needed. And if the role-based filtering is needed, what's `OrchestrationComplete` solving? We should pick one approach.

3. Duplicate `DataT` definition. The diff for `_events.py` has the `TypeVar` import block and `DataT = TypeVar(...)` defined twice — looks like a copy-paste bug.

4. `messages: list[Any]` on `OrchestrationComplete`. Should this be `list[Message]` for type safety?

5. Breaking change for direct workflow consumers. Anyone consuming workflow events directly (not via `.as_agent()`) now gets `OrchestrationComplete` instead of `list[Message]`. All the test changes in group_chat, handoff, and sequential confirm this. Should this have the `[BREAKING]` prefix?

6. Should tool messages reach the user? The blocklist (`!= "user"`) lets tool-role messages through to user-facing output. Are tool results intended to be surfaced, or are they internal artifacts?
